### PR TITLE
Include request headers in client.Info

### LINKED ISSUE #2401: Access headers in processor
**Is your feature request related to a problem? Please describe.**

I would like to access HTTP headers in the processor pipeline and add attribute to span/resource - e.g. tenant ID.

**Describe the solution you'd like**

Read the headers from the context object in a processor. The headers which should be injected into the context could be specified in the receiver config. 

**Describe alternatives you've considered**

A custom implementation of Zipkin/OTLP receiver that does that. 

**Additional context**

Related issue https://github.com/open-telemetry/opentelemetry-collector/issues/2101

### PR DESCRIPTION
**Pull request by disq**
**Description:** Exposes HTTP headers from the originating request in the `client.Info` struct, to be passed on to processors.

This was suggested in the contrib issue https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/6700
I tried to keep it in-line with the `Header` field in `http.Request` (and grpc `metadata.MD`) as possible (so I just used `map[string][]string` in an immutable struct with a getter)

Closes #2401

### CODE REVIEW

**Review by jpkrohling — COMMENTED** [review]
This looks like a good start, but we are missing:

- A similar change to the configgrpc, adding the metadata to the Header. This might be helpful: https://pkg.go.dev/google.golang.org/grpc/metadata#MD
- Tests

**Review by jpkrohling — APPROVED** [review]
LGTM

File: config/confighttp/clientinfohandler.go
Comment (bogdandrutu) [inline config/confighttp/clientinfohandler.go:49]: I think you need to "clone" the headers, otherwise it is not guaranteed that they are not changed/modified after the Handler func finishes. Because of that I think we should have this configured by the user.

**Review by bogdandrutu — CHANGES_REQUESTED** [review]

File: config/confighttp/clientinfohandler.go
Comment (disq) [reply config/confighttp/clientinfohandler.go]: @bogdandrutu Makes sense, but I'm not sure how to make this configurable. Configurable `clientInfoHandler`? It'll eventually circle back to an auth-like extension in the original contrib PR.

**Comment by bogdandrutu** [comment]
@disq @jpkrohling can we document better the motivation of this?

**Comment by disq** [comment]
@bogdandrutu Since I'm running the collector behind a load balancer, my motivation is actually same one as having the `net.Addr` in the `client.Info`: having access to the "real" client IP in the processor.

There are other motivations such as running a multi-tenant infra to distinguish between tenants without having to inject something in the actual OTLP data, which is documented in #2401

**Comment by jpkrohling** [comment]
>@disq @jpkrohling can we document better the motivation of this?

I linked #2401 to this, which should provide enough background on the request.

File: config/confighttp/clientinfohandler.go
Comment (jpkrohling) [inline config/confighttp/clientinfohandler.go:31]:
```suggestion
	// include client metadata or not
```

File: config/confighttp/confighttp.go
Comment (jpkrohling) [inline config/confighttp/confighttp.go:203]:
```suggestion
	// Include propagates the client metadata from the incoming requests to the downstream consumers
```

File: config/configgrpc/configgrpc.go
Comment (jpkrohling) [inline config/configgrpc/configgrpc.go:173]:
```suggestion
	// Include propagates the incoming connection's metadata to downstream consumers.
```

**Review by jpkrohling — COMMENTED** [review]
Apart from a typo and improved godoc for the new property, this looks good to me.

**Review by jcchavezs — APPROVED** [review]

**Review by jpkrohling — APPROVED** [review]

File: config/confighttp/confighttp.go
Comment (bogdandrutu) [inline config/confighttp/confighttp.go:204]: Comments in golang should start with the name of the variable.

Comment (disq) [reply config/confighttp/confighttp.go]: Fixed!

**Review by bogdandrutu — APPROVED** [review]

**Comment by jpkrohling** [comment]
There's a test failure here:

```
=== RUN   TestContextWithClient/existing_client_with_metadata,_no_metadata_processing
    configgrpc_test.go:660: 
        	Error Trace:	configgrpc_test.go:660
        	Error:      	Not equal: 
        	            	expected: client.Info{Addr:net.Addr(nil), Auth:client.AuthData(nil), Metadata:map[string][]string(nil)}
        	            	actual  : client.Info{Addr:net.Addr(nil), Auth:client.AuthData(nil), Metadata:map[string][]string{"test-metadata-key":[]string{"test-value"}}}
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -3,3 +3,7 @@
        	            	  Auth: (client.AuthData) <nil>,
        	            	- Metadata: (map[string][]string) <nil>
        	            	+ Metadata: (map[string][]string) (len=1) {
        	            	+  (string) (len=17) "test-metadata-key": ([]string) (len=1) {
        	            	+   (string) (len=10) "test-value"
        	            	+  }
        	            	+ }
        	            	 }
        	Test:       	TestContextWithClient/existing_client_with_metadata,_no_metadata_processing
```

**Review by codeboten — APPROVED** [review]

**Comment by jpkrohling** [comment]
@disq, you have enough approvals :-) Could you fix the changelog conflict? Once this is done, this PR will be merged.

**Comment by disq** [comment]
> @disq, you have enough approvals :-) Could you fix the changelog conflict? Once this is done, this PR will be merged.

Another conflict? Fixed

**Comment by jpkrohling** [comment]
> Another conflict? Fixed

Sorry about that, we do need to find a better solution for the frequent conflicts involving the changelog.

File: client/client.go
Comment (bogdandrutu) [inline client/client.go:111]: Sorry that I missed this. But because of map/slice usage here, this makes Info mutable which needs to be thread-safe and will complicate lots of things.

@jpkrohling we discussed about this and initially we thought that returning a copy of the Info will be good, but looks like we literally have to provide getters to ensure immutability.

**Review by bogdandrutu — CHANGES_REQUESTED** [review]
Sorry I have to blocked this again, but I missed it. We cannot make Info (a.k.a an object in the context) mutable unless is thread-safe because the context is shared between multiple gorutines potentially.

File: client/client.go
Comment (disq) [reply client/client.go:111]: ouch. It'd have to be unexported with getter _and_ setter, because unexported fields won't be accessible from `config*.contextWithClient` since it's in another package... Info being a non-pointer also makes things difficult, so more code surface needs to change to use `*client.Info`, at least on the set-path.

Comment (disq) [reply client/client.go:111]: I think adding an immutable `*Metadata` fixes a lot of things, trying that route...

Comment (bogdandrutu) [reply client/client.go:111]: Let's see, maybe just add that API without changing all the code to take a look.

Comment (disq) [reply client/client.go]: I don't think we need a mutex as map reads should be thread safe, and we return a copy of the result.

File: client/client.go
Comment (bogdandrutu) [inline client/client.go:111]: Cannot use pointer here because this allows me to do:

```
*(FromContext(ctx).Metadata) = myMetadata
```

This will change the value in the context.

Comment (disq) [reply client/client.go]: @bogdandrutu correct, fixed

File: client/client.go
Comment (trasc) [inline client/client.go:162]:
```suggestion
	vals, found := m.data[key]
	
	if !found {
		return []string{}
	}
	
	ret := make([]string, len(vals))
	copy(ret, vals)

	return ret
```
The main issue is that copy is limited by len(dest) not cap(dest), an the current implementation will always return an empty slice.

**Review by trasc — CHANGES_REQUESTED** [review]

File: client/client.go
Comment (disq) [reply client/client.go]: Thanks @trasc! fixed

File: client/client.go
Comment (bogdandrutu) [inline client/client.go:162]: I think we should check `if len(vals) == 0 return nil`

File: config/configgrpc/configgrpc.go
Comment (bogdandrutu) [inline config/configgrpc/configgrpc.go:157]: Do we want users to be able to select which "metadata" to copy from the request? Maybe not now, but we can be safe and mark this experimental for the moment.

**Review by bogdandrutu — APPROVED** [review]

**Review by trasc — APPROVED** [review]

File: client/client.go
Comment (disq) [reply client/client.go:162]: works as-is per test, what's the difference here?

File: config/configgrpc/configgrpc.go
Comment (disq) [reply config/configgrpc/configgrpc.go:157]: Certainly doable, blindly copying all of it could be bad for memory if the request is decorated with large metadata? But that would also mean we'd have to roll our own `md.Copy()` or `header.Clone()` routines, as copying-first-then-filtering would be pointless in terms of memory.

File: client/client.go
Comment (bogdandrutu) [reply client/client.go:162]: You can write your own benchmarks as well:

```
cpu: Intel(R) Core(TM) i9-9880H CPU @ 2.30GHz
BenchmarkGetIf
BenchmarkGetIf-16          	298160968	         3.912 ns/op	       0 B/op	       0 allocs/op
BenchmarkGetIfExists
BenchmarkGetIfExists-16    	40360434	        27.66 ns/op	      16 B/op	       1 allocs/op
BenchmarkGet
BenchmarkGet-16            	128395258	         9.024 ns/op	       0 B/op	       0 allocs/op
BenchmarkGetExists
BenchmarkGetExists-16      	41000848	        28.28 ns/op	      16 B/op	       1 allocs/op

var data = map[string][]string{
	"key": {"value"},
}

func get(key string) []string {
	vals := data[key]
	ret := make([]string, len(vals))
	copy(ret, vals)

	return ret
}

func getIf(key string) []string {
	vals := data[key]
	if len(vals) == 0 {
		return nil
	}
	ret := make([]string, len(vals))
	copy(ret, vals)

	return ret
}

func BenchmarkGetIf(b *testing.B) {
	b.ResetTimer()
	b.ReportAllocs()
	for n := 0; n < b.N; n++ {
		if len(getIf("test")) != 0 {
			b.Fail()
		}
	}
}

func BenchmarkGetIfExists(b *testing.B) {
	b.ResetTimer()
	b.ReportAllocs()
	for n := 0; n < b.N; n++ {
		if len(getIf("key")) == 0 {
			b.Fail()
		}
	}
}

func BenchmarkGet(b *testing.B) {
	b.ResetTimer()
	b.ReportAllocs()
	for n := 0; n < b.N; n++ {
		if len(get("test")) != 0 {
			b.Fail()
		}
	}
}

func BenchmarkGetExists(b *testing.B) {
	b.ResetTimer()
	b.ReportAllocs()
	for n := 0; n < b.N; n++ {
		if len(get("key")) == 0 {
			b.Fail()
		}
	}
}
```

Comment (disq) [reply client/client.go]: Added, thanks

File: config/configgrpc/configgrpc.go
Comment (bogdandrutu) [reply config/configgrpc/configgrpc.go:157]: That's why I recommend just adding a comment that this is experimental, and we should look a bit into how/if we want this.

Comment (disq) [reply config/configgrpc/configgrpc.go]: @bogdandrutu added

**Review by bogdandrutu — APPROVED** [review]

**Pull request merged**
